Add support for Ruby 3.2 Data Class

egiurleo opened the pull request:
Ruby 3.2 introduced the [`Data` class](https://bugs.ruby-lang.org/issues/16122), which is an immutable data object similar to a `Struct`. This PR adds RBIs for the new `Data` class as well as a rewriter to create correct types for the initializer and add getter methods for the fields on a `Data` subclass.

### Motivation
These changes allow Sorbet users to get valuable typing information about their `Data` classes in the same way they do for `Struct`s.

### Test plan
See included automated tests.

egiurleo commented on rewriter/Data.cc:
```diff
@@ -0,0 +1,136 @@
+            }
+        }
+
+        // TODO(jez) Use Loc::adjust here
```
This file started out as a copy+paste of the [Struct rewriter](https://github.com/sorbet/sorbet/blob/master/rewriter/Struct.cc#L129), which has a few comments that we kept in the file. I'm happy to remove them if folks would prefer.

egiurleo commented on test/testdata/rbi/data.rb:
```diff
@@ -0,0 +1,79 @@
+  point.y
+  T.assert_type!(point, Point)
+
+  point = Point[1, 2]
+  point = Point[x: 1, y: 2]
```
We didn't handle this type of initializer in the writer because it's not handled for `Struct`, but we could try adding it in this or another PR if desired.

jez commented on core/GlobalState.cc:
```diff
@@ -630,6 +630,9 @@ void GlobalState::initEmpty() {
     klass = enterClassSymbol(Loc::none(), Symbols::T_Types(), core::Names::Constants::Base());
     ENFORCE(klass == Symbols::T_Types_Base());
 
+    klass = synthesizeClass(core::Names::Constants::Data());
+    ENFORCE(klass == Symbols::Data());
```
synthesizeClass is a bit low level. Can we use `enterClassSymbol` instead?

jez commented on rewriter/Data.cc:
```diff
@@ -0,0 +1,136 @@
+        return empty;
+    }
+
+    if (!ast::MK::isRootScope(recv->scope) || recv->cnst != core::Symbols::Data().data(ctx)->name ||
```
let's just say

```suggestion
    if (!ast::MK::isRootScope(recv->scope) || recv->cnst != core::Constants::Names::Data() ||
```

jez commented on rewriter/Data.cc:
```diff
@@ -0,0 +1,136 @@
+            }
+        }
+
+        // TODO(jez) Use Loc::adjust here
+        if (symLoc.exists() && absl::StartsWith(ctx.locAt(symLoc).source(ctx).value(), ":")) {
```
```suggestion
        if (symLoc.exists() && ctx.locAt(symLoc).adjustLen(ctx, 0, 1).source(ctx) == ":") {
```

jez commented on rewriter/Data.cc:
```diff
@@ -0,0 +1,136 @@
+
+        // TODO(jez) Use Loc::adjust here
+        if (symLoc.exists() && absl::StartsWith(ctx.locAt(symLoc).source(ctx).value(), ":")) {
+            symLoc = core::LocOffsets{symLoc.beginPos() + 1, symLoc.endPos()};
```
```suggestion
            symLoc = symLoc.adjust(ctx, 1, 0);
```

jez commented:
I have a couple small requests, but other than those this looks good to me.

neilparikh commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_NXrcZXQQkrj0W8
→ https://go/builds/bui_NXrcCXh547O6E9

egiurleo commented:
@jez Thanks for the review! I think I addressed your comments ([see the changes here](https://github.com/sorbet/sorbet/compare/0f39548eb6b02293019ef8f785b454ad1f8b4a4e..11829897cb24fe865086fdb2ad0788a3e5b3f0db)).
